src/libraries/System.Diagnostics.Process/src/System/Diagnostics/ProcessManager.Win32.cs

```diff
@@ -259,31 +256,32 @@ internal static ProcessInfo[] GetProcessInfos(Predicate<int>? processIdFilter =
             // Start with the default buffer size.
             int bufferSize = DefaultCachedBufferSize;
 
-            // Get the cached buffer.
-            long[]? buffer = Interlocked.Exchange(ref CachedBuffer, null);
-
-            try
+            while (true)
```

Comment:
The original code was effectively caching the buffer size between invocations. It does not happen anymore so we may always need an extra iteration before we arrive at the right buffer size on machines with a lot processes now.

Also, ArrayPool has a cut-off at 2MB currently and won't cache anything beyond that size. So this change can result into a lot worse performance on machines with a lot of processes compare to the current implementation.

Reply:
I think we should keep the special local cache, but store it in a WeakHandle so that the GC has a chance to collect if it is not used.

Reply:
> It does not happen anymore

great catch, I've added it back

> Also, ArrayPool has a cut-off at 2MB currently and won't cache anything beyond that size. So this change can result into a lot worse performance on machines with a lot of processes compare to the current implementation.

I've tested it and on my machine with +-150 processes running the API needs 0.6MB. So we should be safe to up to 500 processes. If there are more than 500 processes, the cost of allocating a bigger array should be small compared to the rest of the work. 

Moreover, I think that it should be an edge case as the current trend in the industry to have small microservices running in containers.

Reply:
> I think we should keep the special local cache, but store it in a WeakHandle so that the GC has a chance to collect if it is not used.

I am not sure if replacing the local cache with `WeakHandle` would not make it worse. From my experience `WeakHandle`-based caches are typically more expensive than allocation as they are used just once and get released with next GC.

Reply:
@jkotas do we have any other custom byte buffer caches in the framework?

FWIW I've provided some benchmark results in the PR description

Reply:
>  think that it should be an edge case as the current trend in the industry to have small microservices running in containers.

This is for Windows. I can easily imagine big first party workloads on Windows hitting this limit.

> the cost of allocating a bigger array should be small compared to the rest of the work.

The problem are secondary effects of allocating the big buffer. For example, this API can be called in background monitoring tasks. When this background monitoring tasks starts producing a lot of Gen2 garbage, the health of the whole process gets effected.

> From my experience WeakHandle-based caches are typically more expensive than allocation as they are used just once and get released with next GC.

That's true for caches of small objects - they will get released with next Gen0 GC. This buffer is huge - it will only get released with next Gen2 GC.

> do we have any other custom byte buffer caches in the framework?

Caching huge buffers is generally problematic. An alternative is to not use ArrayPool at all here, and just use unmanaged memory allocator, without any caches. It would eliminate the need for the alignment tricks. The memory returned by `Marshal.AllocHGlobal` is guaranteed to have sufficient alignment.

Reply:
So perhaps we should combine the two ideas? Rent small arrays to avoid managed allocations, allocate unmanaged memory for bigger buffers?

@jkotas PTAL at the updated code and benchmark numbers in the PR description